pkg/kb_agent/httpserver/server.go

```diff
@@ -0,0 +1,163 @@
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	fasthttprouter "github.com/fasthttp/router"
+	"github.com/valyala/fasthttp"
+)
+
+// Server is an interface for the kb-agent HTTP server.
+type Server interface {
+	io.Closer
+	Router() fasthttp.RequestHandler
+	StartNonBlocking() error
+}
+
+type server struct {
+	config    Config
+	endpoints []Endpoint
+	servers   []*fasthttp.Server
+}
+
+// NewServer returns a new HTTP server.
+func NewServer() Server {
+	return &server{
+		endpoints: Endpoints(),
+		config:    config,
+	}
+}
+
+// StartNonBlocking starts a new server in a goroutine.
+func (s *server) StartNonBlocking() error {
+	logger.Info("Starting HTTP Server")
+	handler := s.Router()
+
+	APILogging := s.config.APILogging
+	if APILogging {
+		handler = s.apiLogger(handler)
+	}
+
+	var listeners []net.Listener
+	if s.config.UnixDomainSocket != "" {
+		socket := fmt.Sprintf("%s/kb_agent.socket", s.config.UnixDomainSocket)
+		l, err := net.Listen("unix", socket)
+		if err != nil {
+			return err
+		}
+		listeners = append(listeners, l)
+	} else {
+		apiListenAddress := s.config.Address
+		l, err := net.Listen("tcp", fmt.Sprintf("%s:%v", apiListenAddress, s.config.Port))
+		if err != nil {
+			logger.Error(err, "listen address", apiListenAddress, "port", s.config.Port)
+		} else {
+			listeners = append(listeners, l)
+		}
+	}
+
+	if len(listeners) == 0 {
+		return errors.New("could not listen on any endpoint")
+	}
+
+	for _, listener := range listeners {
+		// customServer is created in a loop because each instance
+		// has a handle on the underlying listener.
+		customServer := &fasthttp.Server{
+			Handler:            handler,
+			MaxRequestBodySize: s.config.MaxRequestBodySize * 1024 * 1024,
+			ReadBufferSize:     s.config.ReadBufferSize * 1024,
+		}
+		s.servers = append(s.servers, customServer)
+
+		go func(l net.Listener) {
+			if err := customServer.Serve(l); err != nil {
+				panic(err)
+			}
+		}(listener)
+	}
+
+	return nil
+}
+
+func (s *server) apiLogger(next fasthttp.RequestHandler) fasthttp.RequestHandler {
+	return func(ctx *fasthttp.RequestCtx) {
+		reqLogger := logger
+		if userAgent := string(ctx.Request.Header.Peek("User-Agent")); userAgent != "" {
+			reqLogger = logger.WithValues("useragent", userAgent)
+		}
+		start := time.Now()
+		path := string(ctx.Path())
+		if path == "/v1.0/checkrole" {
+			// do not log for checkrole
+			next(ctx)
+			return
+		}
+
+		reqLogger.Info("HTTP API Called", "method", string(ctx.Method()), "path", path)
+		next(ctx)
+		elapsed := float64(time.Since(start) / time.Millisecond)
+		reqLogger.Info("HTTP API Called", "status code", ctx.Response.StatusCode(), "cost", elapsed)
+	}
+}
+
+func (s *server) Router() fasthttp.RequestHandler {
+	router := s.getRouter(s.endpoints)
+
+	return router.Handler
+}
+
+func (s *server) getRouter(endpoints []Endpoint) *fasthttprouter.Router {
+	router := fasthttprouter.New()
+	for _, e := range endpoints {
+		path := fmt.Sprintf("/%s/%s", e.Version, e.Route)
```

Comment:
It seems there is no check performed using e.Version?

Reply:
Yes, e.Version represents the current version of the kb-agent API service, which is currently v1.0. the ChatGPT explanation for use of the version:
![image](https://github.com/apecloud/kubeblocks/assets/5369071/1db3b7b9-b752-43fd-9e96-0085b098ae3e)

Reply:
There should be have some compatibility design principles here. For example:
- should a kb-agent of v1.0 process a message of v1.x ?
- should a kb-agent of v1.0 process a message of v2.0 ?
- should a kb-agent of v2.0 process a message of v1.x ?

Besides, the current version of kb-agent should be defined in a constant which will be printed to log when kb-agent starts up and the version is checked according to above compatibility logics.

Reply:
yes, the version is introduced to allow breaking changes. If kb-agent is upgraded to v2.0, the v1.0 API will be maintained for compatibility until all legacy accesses are removed.